Fix build for MSVC (Windows x64) and fix all warnings reported by MSVC

lakor64 opened the pull request:
The following commit fixes Pyston's forked CPython interpreter for MSVC.
It also fixes all the warnings reported by MSVC on the default level (which is -W3).
I haven't currently tested with the AOT macro enabled (is there any other macro that should be enabled as well?) as I though just pyston optimizations would be a start.

I've also updated some branding settings in the props so the built dll would be pyston23.dll and not python38.dll.

I haven't tested the C modules (such as _ctypes because of some dep compilation error which I didn't bother to fix now, if there's a fix required I might send a new PR for dependency update on Windows etc...).

Currently, python.exe properly builds and executes on both debug and release with pyston speedups enabled, a screenshot can be seen here: https://i.imgur.com/2Le4K1N.png.

I haven't done extensive testing expect a print and sys import, so I'm not sure how stable it would be.

The project can be built by opening PCbuild\pcbuild.sln with Visual Studio, I personally used VS2022.

NOTE: Some other branding changes would be the resource/rc files, regarding of branding, should the interpreter be built as python3.exe or pyston2.exe?

(Some macros were backported from CPython 3.12 including exports.h)

lakor64 committed:
Fix build for MSVC (Windows x64) and fix all warnings reported by MSVC

kmod commented:
This is amazing, thank you!

As to your question, I think it should be pyston3.exe -- I believe people expect the version number to be the version of the Python language, not the version of the interpreter.

lakor64 commented:
Regarding the branding also, ![image](https://user-images.githubusercontent.com/55551484/205493632-991290ea-7f0b-4aa0-9076-af03f5b858ac.png) should the name of this manifest info be modified accordingly? eg: Python Core (etc..)

lakor64 committed:
Remove inclusion of vcpkg, implement lite/simple builds (incomplete), adjustments macros for AOT and debugging, adjust python exe/dll branding

lakor64 committed:
Remove automatic inclusion of python38, and adjust compiler macros to pyston needs

lakor64 committed:
Fix other files for windows and remove masm usage

lakor64 committed:
Disable AOT and fix killing istances of pyston

lakor64 commented:
I nthe next commits I've added some preliminary work for AOT, as well as fixing _ctypes compilation.
Judging from the Makefile that is left at the root (I've left some notes on discord as well but the discord doesn't seem active, should I write such notes here?).
The PYSTON_LITE switch doesn't seem to be working here.
My idea for the whole AOT enabled Pyston (which I would send some other PR for merging Windows support into pystol and nitrous later) would be the following:
0. Create a SimpleRelease/SimpleDebug configurations that generates a simple AOT disabled pyston with speedup enabled (what this commits achieves so far)
1. Run pcbuild.sln (or another sln) with the configuration SimpleRelease, this should generate a simple (pyston-simple.exe) interpreter without AOT, nitrous and pystol
2. Run the AOT generation tool with debug and release options
3. Attempt to build pyston with AOT enabled

I'm not sure if the makefile goes into other steps of the compilation, but this seems to be the steps that are executed for having an AOT pyston.

kmod commented:
> Regarding the branding also, ![image](https://user-images.githubusercontent.com/55551484/205493632-991290ea-7f0b-4aa0-9076-af03f5b858ac.png) should the name of this manifest info be modified accordingly? eg: Python Core (etc..) (First line is File description the second is the company and so on) (src is at https://github.com/pyston/pyston/blob/pyston_main/PC/python_nt.rc)

Why don't we make it "Accelerated Python" and "The Pyston Team"

kmod commented:
> I nthe next commits I've added some preliminary work for AOT, as well as fixing _ctypes compilation. (and possibly for anyone who uses vcpkg in the system) The PYSTON_LITE switch doesn't seem to be working here while I left two compilations for it (LiteDebug and LiteRelease) they doesn't seem to build properly and I'm unsure if it's the same case for Linux as well. Judging from the Makefile that is left at the root (I've left some notes on discord as well but the discord doesn't seem active, should I write such notes here?). My idea for the whole AOT enabled Pyston (which I would send some other PR for merging Windows support into pystol and nitrous later) would be the following: 0. Compile clang for WIndows (for generating ILCode) and LLVM (In theory clang for windows could be used but requires extra thinkering for Visual Studio so I would having it genering a MSVC enabled LLVM) (bolt will be skipped as it doesn't work with PE afaik)
> 
> 1. Run pcbuild.sln (or another sln) with the configuration SimpleRelease, this should generate a simple (pyston-simple.exe) interpreter without AOT, nitrous and pystol
> 2. Run the AOT generation tool with debug and release options
> 3. Attempt to build pyston with AOT enabled
> 
> I'm not sure if the makefile goes into other steps of the compilation and I haven't completely figure out where aot.h is generated (aot_gen does produce some files)

I know nothing about compilation on Windows so I'm sorry that I can't be much help with that angle :/

pyston-lite has a completely different build system through a standard setup.py -- it should be enough to go into pyston/pyston_lite and do `python3 setup.py wheel`, assuming that works on windows. It will set the right environment variables and everything for the build

I think it's easiest to keep the discussion about this change here vs discord -- also I don't get email notifications from discord so I just check it once a week, whereas I get notifications from github and check it more actively.

And sorry, yeah the build system is complex and not well-documented. I believe aot.h is generated by aot_gen.py, but I often forget which files are generated there and have to run the makefile step by step to redetermine it.

lakor64 committed:
Update branding

lakor64 commented:
> > I nthe next commits I've added some preliminary work for AOT, as well as fixing _ctypes compilation. (and possibly for anyone who uses vcpkg in the system) The PYSTON_LITE switch doesn't seem to be working here while I left two compilations for it (LiteDebug and LiteRelease) they doesn't seem to build properly and I'm unsure if it's the same case for Linux as well. Judging from the Makefile that is left at the root (I've left some notes on discord as well but the discord doesn't seem active, should I write such notes here?). My idea for the whole AOT enabled Pyston (which I would send some other PR for merging Windows support into pystol and nitrous later) would be the following: 0. Compile clang for WIndows (for generating ILCode) and LLVM (In theory clang for windows could be used but requires extra thinkering for Visual Studio so I would having it genering a MSVC enabled LLVM) (bolt will be skipped as it doesn't work with PE afaik)
> > 
> > 1. Run pcbuild.sln (or another sln) with the configuration SimpleRelease, this should generate a simple (pyston-simple.exe) interpreter without AOT, nitrous and pystol
> > 2. Run the AOT generation tool with debug and release options
> > 3. Attempt to build pyston with AOT enabled
> > 
> > I'm not sure if the makefile goes into other steps of the compilation and I haven't completely figure out where aot.h is generated (aot_gen does produce some files)
> 
> I know nothing about compilation on Windows so I'm sorry that I can't be much help with that angle :/
> 
> pyston-lite has a completely different build system through a standard setup.py -- it should be enough to go into pyston/pyston_lite and do `python3 setup.py wheel`, assuming that works on windows. It will set the right environment variables and everything for the build
> 
> I think it's easiest to keep the discussion about this change here vs discord -- also I don't get email notifications from discord so I just check it once a week, whereas I get notifications from github and check it more actively.
> 
> And sorry, yeah the build system is complex and not well-documented. I believe aot.h is generated by aot_gen.py, but I often forget which files are generated there and have to run the makefile step by step to redetermine it.

Thanks for your reply, I'll take a look with a Linux instance then to further investigate the build process and stipulate a way to build Pyston with JIT (or even Pyston lite).

I've updated the branding in the last commit, if there is something else to change you can let me know.

In regards of this PR, do you prefer having all the changes in one single PR or proceed to review this PR (which includes only basic support for windows compilation with pyston speedups) or wait for future PRs for Pystol and so on?

kmod commented:
Yes we can definitely merge this PR before the rest of the work, but I'm seeing a lot of build and test failures

(We have some known test failures on the 3.10 configuration, but all the other ones seem new to this pr.)

WoolenWang commented:
when will this pr fix all test and merge into ? hope to have the windows pyston .

lakor64 commented:
Hello, bumping this issue again, I  will check CI issues in this days on my local machine (been having some issues in building pyston from a linux machine unreleated to this) and ask for a review again.
Answering @WoolenWang this PR won't currently allow you to have full support of Pyston on Windows (eg: AOT is missing), but it would be a first step torwards full pyston compatibility with Windows.

kmod commented:
Hi @lakor64 I thought I had responded to your questions -- is there something you're waiting for from me?

lakor64 commented:
> Hi @lakor64 I thought I had responded to your questions -- is there something you're waiting for from me?

Hello, I think you answered all my questions, altrough I fail to have such issue in my own computer (the one reported by the CI). I wonder if I keep pushing to this branch does the CI runs again? I don't know about CI in github

kmod commented:
Yes the CI will run now if you push to this branch -- it didn't originally as a security precaution, until I enabled it. However I'd recommend using a different branch and using the CI from your own repo, because pushing to this branch will notify the people on the thread.

I'm not sure why you haven't had any [CI actions run](https://github.com/lakor64/pyston/actions) so far in your repo -- you might have to enable something. You can see in [my repo](https://github.com/kmod/pyston/actions) that it regularly runs the CI on changes I commit to my personal branches.

Do you have a Linux machine to test things on? That is probably faster than testing via CI.

lakor64 commented:
Thank you for your input, I have enabled in my repo github actions, while for the Linux computer, I did setup a local Linux machine to test Pyston implementation but I don't seem to get right on how to build it, I tried to follow the CI and passed `make test` but strangely it doesn't seem to build pyston properly, apperently shortly after I hit build the compilation just ends, I belive after it did compilation of bolt. I had tried to run the make command multiple time but without much of success. I will attempt to try it later with the use of the CI.

WoolenWang commented:
> 

still thankful for this outstanding pull
